Add horizontal pod autoscaler in monitoring chapter (#738)
SUSE/doc-caasp: adoc/admin-monitoring-horizontal-pod-autoscaler.adoc

= Horizontal Pod Autoscaler

Horizontal Pod Autoscaler (HPA) allows to increase/decrease the number of pods in a replication controller, deployment, replica set or stateful set, based on the metrics collected from pods.

In order to leverage HPA, {skuba} now supports an addon `metrics-server` installed link:https://github.com/kubernetes-sigs/metrics-server[metrics-server] into {kube} cluster.

After that, HPA fetches metrics from the aggregated API `metrics.k8s.io` and according to the criteria user configuration to determine whether to increase/decrease the scale of a replication controller, deployment, replica set or stateful set.

The HPA `metrics.target.type` could be one of:

> - Utilization: the value returned from the metrics server API is calculated to average CPU utilization and compare to the `metrics.target.averageUtilization`.

Reviewer: ```suggestion
- Utilization: the value returned from the metrics server API is calculated to average resource utilization across all relevant pods, then compare to the `metrics.target.averageUtilization`.
```

> - AverageValue: the value returned from the metrics server API is divided by the number of pods and compared to the `metrics.target.averageValue`.

Reviewer: ```suggestion
- AverageValue: the value returned from the metrics server API is divided by the number of all relevant pods, then compared to the `metrics.target.averageValue`.
```

- Value: the value returned from the metrics server API is directly compared to the `metrics.target.value`.

[NOTE]
====

> The metrics are CPU and memory of a pod and node.

Reviewer: ```suggestion
The metrics supported by `metrics-server` are CPU and memory of a pod and node.
```

** HPA by CPU

*** Create HPA by average CPU utilization
+
```
# deployment
kubectl autoscale deployment <deployment-name> \
    --min=<min_replicas_number> \
    --max=<max_replicas_number> \
    --cpu-percent=<percent>

# replication controller
kubectl autoscale replicationcontrollers <replicationcontrollers-name> \
    --min=<min_replicas_number> \
    --max=<max_replicas_number> \
    --cpu-percent=<percent>
```
For example:
+
```
kubectl autoscale deployment example \
    --name=avg-cpu-util \
    --min=1 \
    --max=10 \
    --cpu-percent=50
```
The example shows autoscaling for the example deployment, the HPA increase  the minimum number of pod to 1 and will increase the pods up to 10 if the avergae CPU utilization of the pods reaches 50%. Ref to link:https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details[kubernetes: horizontal pod autoscale alrorithm details] for the algorithm details.
+
Check the current status of the HPA:
+
```
kubectl get hpa

NAME           REFERENCE            TARGETS   MINPODS   MAXPODS   REPLICAS   AGE
avg-cpu-util   Deployment/example   0%/50%    1         10        1          2m4s
```
+
[NOTE]
====

> HPA calculates pod cpu utilization as: total cpu usage of all containers / total cpu requests. For example:

Reviewer: ```suggestion
HPA calculates pod CPU utilization as: total CPU usage of all containers / total CPU requests. For example:
```

etcd-master000                      21m          584Mi
kube-apiserver-master000            20m          325Mi
kube-controller-manager-master000   6m           105Mi
kube-proxy-x2965                    0m           24Mi
kube-proxy-x9zlv                    0m           19Mi
kube-scheduler-master000            2m           46Mi
kured-45rc2                         1m           25Mi
kured-cptk4                         0m           25Mi
metrics-server-79b8658cd7-gjvhs     1m           21Mi
oidc-dex-55fc689dc-f6cfg            1m           20Mi
oidc-gangway-7b7fbbdbdf-85p6t       1m           18Mi
----
+
[NOTE]
====
The option flag `--sort-by=cpu`/`--sort-by=memory` has an sorting issue at the moment. It will be fixed in the future.
====

=== Using Horizontal Pod Autoscaler (HPA)

You can set the HPA to scale according to various metrics.
These include *average CPU utilization*, *average CPU value*, *average memory utilization* and *average memory value*. The following sections show the recommended configuration for each of the aforementioned options.

==== Creating an HPA Using Average CPU Utilization

The following code is an example of what this type of HPA can look like:

TODO::

>   Where does this configuration go? (which config file?)

Reviewer 2: @jenting ah, sorry, I marked the question as TODO. I am not sure where the following configuration goes.

etcd-master000                      21m          584Mi
kube-apiserver-master000            20m          325Mi
kube-controller-manager-master000   6m           105Mi
kube-proxy-x2965                    0m           24Mi
kube-proxy-x9zlv                    0m           19Mi
kube-scheduler-master000            2m           46Mi
kured-45rc2                         1m           25Mi
kured-cptk4                         0m           25Mi
metrics-server-79b8658cd7-gjvhs     1m           21Mi
oidc-dex-55fc689dc-f6cfg            1m           20Mi
oidc-gangway-7b7fbbdbdf-85p6t       1m           18Mi
----
+
[NOTE]
====
The option flag `--sort-by=cpu`/`--sort-by=memory` has an sorting issue at the moment. It will be fixed in the future.
====

=== Using Horizontal Pod Autoscaler (HPA)

You can set the HPA to scale according to various metrics.
These include *average CPU utilization*, *average CPU value*, *average memory utilization* and *average memory value*. The following sections show the recommended configuration for each of the aforementioned options.

==== Creating an HPA Using Average CPU Utilization

The following code is an example of what this type of HPA can look like:

TODO::

>   Where does this configuration go? (which config file?)

Author: Sorry I didn't catch this question point. But I'd try the guess your question.

It would create autoscale rule to Kubernetes cluster based on CPU percentage, with min&max replicas number and monitoring the name of Deployment. If the deployment name CPU percentage above/below the criteria, the Kubernetes would help us increase/decrease the number of deployment pods.

Reviewer 2: Hmm, I don't think we understand each other. :) I am asking about the code that is just below my question (now I understand what it is supposed to do), but I don't understand where it gets executed. Or in other words what steps do you take exactly, as a caasp admin, to get from this code below to a point where you have the working autoscale rule.

Author: I think maybe removing these blocks make it easier to read. WDYT?

```
# deployment
kubectl autoscale deployment <deployment-name> \
    --min=<min_replicas_number> \
    --max=<max_replicas_number> \
    --cpu-percent=<percent>

# replication controller
kubectl autoscale replicationcontrollers <replicationcontrollers-name> \
    --min=<min_replicas_number> \
    --max=<max_replicas_number> \
    --cpu-percent=<percent>
```

Author: For a CaaSP admin, for example, I deploy a server Deployment _\<deployment-name\>_ and I want Kubernetes to auto increase pod number when burst requests coming; decrease pod number when no bust requests coming.

Then, I would like to set min pod and max pod, and based on what criteria (CPU or memory) to determine when to increase or decrease pod number. Therefore, the CaaSP admin would set up autoscale rules into Kubernetes clusters.
```
kubectl autoscale deployment <deployment-name> \
    --min=<min_replicas_number> \
    --max=<max_replicas_number> \
   --cpu-percent=<percent>
```

Then, get horizontal pod autoscaler (HPA) to get/list/watch the rule CaaSP admin setting
```
kubectl get hpa
```

Reviewer 2: So you would run that code on the admin node/control plane?

Author: Admin node or user local machine.

The user needs kubeconfig file with corresponding RBAC permission to allows set up autoscale rules into K8s cluster.

Reviewer 2: Ok, I think that's what I wanted to know. Thanks. :)

oidc-dex   Deployment/oidc-dex   0%/50%          1         10        3          115s
```

[NOTE]
====
To calculate pod CPU utilization HPA divides the total CPU usage of all containers by the total number of CPU requests:

POD CPU UTILIZATION = TOTAL CPU USAGE OF ALL CONTAINERS / NUMBER OF CPU REQUESTS

For example:

- Container1 requests 0.5 CPU and uses 0 CPU.
- Container2 requests 1 CPU and uses 2 CPU.

The CPU utilization will be (0+2)/(0.5+1)*100 (%)=133 (%)

If a replication controller, deployment, replica set or stateful set does not specify the CPU request, the output of `kubectl get hpa` TARGETS will be unknown.
====

==== Creating an HPA Using the Average CPU Value

. Create a yaml manifest file `hpa-avg-cpu-value.yaml` with the following content:
+
```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: avg-cpu-value // <1>
  namespace: kube-system // <2>
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment // <3>
    name: example // <4>
  minReplicas: 1 // <5>
  maxReplicas: 10 // <6>
  metrics:
  - type: Resource
    resource:
      name: cpu
      target:
        type: AverageValue

>         averageValue: 500Mi <7>

Reviewer 3: ```suggestion
        averageValue: 500Mi // <7>
```

Author: nice catch! thx.

<3> Specifies the kind of object to scale (a replication controller, deployment, replica set or stateful set).
<4> Specifies the name of the object to scale.
<5> Specifies the minimum number of replicas.
<6> Specifies the maximum number of replicas.
<7> The average value of the requested CPU that each pod uses.

. Apply the yaml manifest by running:
+
```
kubectl apply -f hpa-avg-cpu-value.yaml
```

. Check the current status of the HPA:
+
```
kubectl get hpa

NAME            REFERENCE               TARGETS    MINPODS   MAXPODS   REPLICAS   AGE
avg-cpu-value   Deployment/php-apache   1m/500Mi   1         10        1          39s
```

==== Creating an HPA Using Average Memory Utilization

. Create a yaml manifest file `hpa-avg-memory-util.yaml` with the following content:
+
```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: avg-memory-util // <1>
  namespace: kube-system // <2>
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment // <3>
    name: example // <4>
  minReplicas: 1 // <5>
  maxReplicas: 10 // <6>
  metrics:
  - type: Resource
    resource:
      name: memory
      target:
        type: Utilization

>         averageUtilization: 50 <7>

Reviewer 3: ```suggestion
        averageUtilization: 50 // <7>
```

. Apply the yaml manifest by running:
+
```
kubectl apply -f hpa-avg-memory-util.yaml
```

. Check the current status of the HPA:
+
```
kubectl get hpa

NAME              REFERENCE            TARGETS          MINPODS   MAXPODS   REPLICAS   AGE
avg-memory-util   Deployment/example   5%/50%           1         10        1          4m54s
```
+
[NOTE]
====
HPA calculates pod memory utilization as: total memory usage of all containers / total memory requests.
If a deployment or replication controller does not specify the memory request, the ouput of `kubectl get hpa` TARGETS is <unknown>.
====

==== Creating an HPA Using Average Memory Value

. Create a yaml manifest file `hpa-avg-memory-value.yaml` with the following content:
+
```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: avg-memory-value // <1>
  namespace: kube-system // <2>
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: Deployment // <3>
    name: example // <4>
  minReplicas: 1 // <5>
  maxReplicas: 10 // <6>
  metrics:
  - type: Resource
    resource:
      name: memory
      target:
        type: AverageValue

>         averageValue: 500Mi <7>

Reviewer 3: ```suggestion
        averageValue: 500Mi // <7>
```
